### PR TITLE
Feature: ORM Driven Character Positions for Issue #166

### PR DESCRIPTION
**Pull request by fantoms**
This huge PR is related to Issue #166. The code has been changed to allow more then one position type. A class was added for Character position types that match the requirements listed in the issue. A secondary Position Type class was added to match the client, provided by @LtRipley36706.

To work with the ORM code introduced during the developer meeting, I've attempted to copy the pseudo 1:1 mapping into `CharacterPosition.cs`. In order to allow for more then one position, the `Character.Position` member was changed to dictionary containing a list of character positions by type. `Player.Position` was changed to `Player.PhysicalPosition` and when the server decides to save the positions to the database, the data is saved through the ORM object in `CharacterPosition.cs`. 

There were a few things left to do, that were not accomplished for Issue #166:
* Hook up /lifestone and /ls commands

Changelog with this PR: 

### 2017-03-31
[fantoms]
* Added debug command to test positions.
* Changed SQL syntax to use BEFORE instead of AFTER for the `positionType` column, in the `character_positions` table.
* Removed the github friendly enums from the `PositionTypes`.
* Changed a log4net debug line.
* Changed the name class name `CharacterPositionType` to `PositionType`
* Reworked `Position.cs` with ORM.

### 2017-03-29
[fantoms]
* Changed the Player.Position to Player.PhysicalPosition to match the Developer meeting video and Issue #166.
* Created `CharacterPostion` and `CharacterPositionType` classes to store the Database types requested in Issue #166.
* Added `CharacterPostion` ORM logic.
* Built out helper extension functions to serve the `StartingLocation`, until we pull from the database. Also included an Invalid location as a default (0).
* Changed the `character_positions` table and Added `positionType` column.
* Set the primary key to include `positionType` column.
* Added `positionType` to required Critera for `character_positions`.
* Added `CharacterPositionInsert` and `CharacterPositionUpdate` to the Character Database.
* Created `CharacterPositionInsert` and `CharacterPositionUpdate` prepared SQL statements by utilizing the `ConstructStatement` function.
* Added the required `positionType` clause to the `CharacterPositionSelect` prepared statement.
* Added a position update statement for updating the player position.
* Reworked GetPosition to include the `CharacterPositionType`.
* Added function for getting a `CharacterPostion` from the database, requires a character Id and `CharacterPositionType`.
* Added logic to get location function to insert a location into the database, if not present.
* Reworked the UpdateCharacter and Update functions to save positions.
* Added functions for inserting new character positions on character creation.
* Added functions for saving/loading the character positions from the database on joining the World.
* Attempted to get log4net working properly and moved some code to logging.
* Added DBDEBUG project constant for debugging the database calls.
* Changed the `CharacterPositionSelect` prepared statement to a `ConstructStatement`.
+ Updates after PR review: +
* Changed the `LoadCharacterPositions` to ORM OUT of the database.
* Changed the introduced @save command to @save-now to avoid conflict
* Changed CharacterPositionType to PositionTypes and merged in Ripley's enums
* Changed PhyiscalPosition to Location
* Removed plural in function name for position types; broke the character loading
* Added database view for all list positions
* Added GetList statement for all positions
* Migrated GetPosition to GetLocation, Seporated Logic, added ORM calls on selecting data
* Moved the Character.Location to Character.CharacterPositions[PositionType.Location] dictionary

### Git Log:
- Changed position to phsysical/logical and added logical positions
- Updated comments on CharacterPositionType
- Added the sql for altering the positions table, to add the positionType
- Added ripleys Position Types
- Added Character Position Database Object
- Renamed Position types to Positions to Match Ripley's original code. Looks better in character, also.
- Changed to the correct function when Executing constructed statements
- Added another debug command
- Added Position update to the Character Data object
- Changed the positiontype to a keyed column in mysql
- DB Updates
- Morning updates - working position inserts at character creation
- Updated constructstatement, positions are saving to the right columns
- Load all available character positions from the database or create a new if missing
- Fixed unused GetCharacterPhysicalPosition
- Made the server use the correct insert function
- Fixed position save
- Cleaned up code and removed the unnecessary SQL statement
- Moved Character Prepared Statement for Selecting position to the ConstructedStatement Get function
- Changed the LoadCharacterPositions to ORM OUT of the database.
- Changed the introduced @save command to @save-now to avoid conflict
- Changed CharacterPositionType to PositionTypes and moved in Ripley's code
- Changed PhyiscalPosition to Location
- Removed plural in function name for position types; broke the character loading
- Added GetList statement for all positions
- Migrated GetPosition to GetLocation, Seporated Logic, added ORM calls on selecting data
- Moved the Character.Location to Character.CharacterPositions[PositionType.Location] dictionary
- Added debug command to test positions
- Migrated enums to ripleys position only - took out the labels from github
- Updated SQL to use BEFORE instead of AFTER
- Changed DBDEBUG line to use log4net
- Began working on position
- Changed the name from CharacterPositionType to PositionType
- Converted Position to rely on floats instead of vectors and began adding ORM logic
- Renamed variables to go inline with code style
- Moved character helper extensions and migrated to Positions
- Migrated positions to orm
- Adding Character Positions view for the GetList query
- Swapped column order for Ripley
- Reworked landblock positions - there is some data loss happening that i would like help locating
- Fixed up comments and updated changelog

### CODE REVIEW

File: Source/ACE.Database/CharacterDatabase.cs [hunk 1]
```diff
@@ -84,13 +88,15 @@ protected override void InitialisePreparedStatements()
...
             AddPreparedStatement(CharacterPreparedStatement.CharacterFriendsSelect, "SELECT cf.`friendId`, c.`name` FROM `character_friends` cf JOIN `character` c ON (cf.`friendId` = c.`guid`) WHERE cf.`id` = ?;", MySqlDbType.UInt32);
 
-            AddPreparedStatement(CharacterPreparedStatement.CharacterPositionInsert, "REPLACE INTO character_position (id, cell, positionX, positionY, positionZ, rotationX, rotationY, rotationZ, rotationW) VALUES (?, ?, ?, ?, ?, ?, ? ,? ,?);", MySqlDbType.UInt32, MySqlDbType.UInt32, MySqlDbType.Float, MySqlDbType.Float, MySqlDbType.Float, MySqlDbType.Float, MySqlDbType.Float, MySqlDbType.Float, MySqlDbType.Float);
+            ConstructStatement(CharacterPreparedStatement.CharacterPositionSelect, typeof(Position), ConstructedStatementType.Get);
+            ConstructStatement(CharacterPreparedStatement.CharacterPositionInsert, typeof(Position), ConstructedStatementType.Insert);
+            ConstructStatement(CharacterPreparedStatement.CharacterPositionUpdate, typeof(Position), ConstructedStatementType.Update);
+            ConstructStatement(CharacterPreparedStatement.CharacterPositionList, typeof(Position), ConstructedStatementType.GetList);
```
Comment (fantoms) [inline Source/ACE.Database/CharacterDatabase.cs]: Changed all of the Position related statements to ConstructStatements.

File: Database/Updates/Character/2017_03_29_character_vw_character_positions.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
+DROP VIEW IF EXISTS `vw_character_positions`;
```
Comment (fantoms) [inline Database/Updates/Character/2017_03_29_character_vw_character_positions.sql]: This script creates a database `view`, used for getting all of the positions for a single character.

File: Source/ACE.Database/CharacterDatabase.cs [hunk 3]
```diff
@@ -260,12 +275,15 @@ public bool IsNameAvailable(string name)
                 c.TemplateOption = result.Read<uint>(0, "templateOption");
                 c.StartArea = result.Read<uint>(0, "startArea");
 
-                c.Position = await this.GetPosition(guid);
+                // Loads the positions into the player object and resets the landlock id
+                LoadCharacterPositions(c);
+                c.Location = c.Positions[PositionType.Location];
+                c.Location.LandblockId = new LandblockId(c.Location.cell);
```
Comment (Mogwai-AC) [inline Source/ACE.Database/CharacterDatabase.cs]: I get why this was necessary and I don't like it either.

File: changelog.md [hunk 4]
```diff
@@ -1,4 +1,45 @@
 # ACEmulator Change Log
+### 2017-03-31
```
Comment (Mogwai-AC) [inline changelog.md]: for future reference, this is way more verbose than i had in mind.

**Pull request merged**